bftclient/src/bft_client.cpp

```diff
@@ -0,0 +1,162 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License"). You may not use
+// this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license
+// terms. Your use of these subcomponents is subject to the terms and conditions of the
+// subcomponent's license, as noted in the LICENSE file.
+
+#include "bftclient/bft_client.h"
+#include "bftengine/ClientMsgs.hpp"
+#include "assertUtils.hpp"
+
+namespace bft::client {
+
+// This function creates a ClientRequestMsg or a ClientPreProcessRequestMsg depending upon config.
+//
+// Since both of these are just instancds of a `ClientRequestMsgHeader` followed by the message
+// data, we construct them here, rather than relying on the type constructors embedded into the
+// bftEngine impl. This allows us to not have to link with the bftengine library, and also allows us
+// to return the messages as vectors with proper RAII based memory management.
+Msg MakeClientMsg(const RequestConfig& config, Msg&& request, bool read_only, uint16_t client_id) {
+  uint8_t flags = read_only ? READ_ONLY_REQ : EMPTY_FLAGS_REQ;
+  if (config.pre_execute) {
+    flags |= PRE_PROCESS_REQ;
+  }
+
+  auto header_size = sizeof(bftEngine::ClientRequestMsgHeader);
+
+  Msg msg(header_size + request.size());
+  bftEngine::ClientRequestMsgHeader* header = reinterpret_cast<bftEngine::ClientRequestMsgHeader*>(msg.data());
```

Comment:
This way of creation messages is dangerous because it does not take into account the variable size of the span context. I am afraid we'll end up creating a separate implementation of messages.

Reply:
I agree it's dangerous, but it's the least worst of the options IMO, given the current message implementation. We shouldn't be adding any new fields before we switch to a better implementation. If we do, we'll have to remember to fix this. I'm pretty sure we will, because tests will probably start failing like crazy. 

Also, thanks for pointing this out, since I actually had a TODO below about taking into account the spanContextSize and forgot to do that. Will fix that.

Reply:
I just updated this to include span context properly in the RequestConfig and put parameters in the right order. How does the new code look?

Reply:
@andrewjstone, thank you for fixing it, IMO, it is a very elegant solution.